src/github.com/getlantern/flashlight/util/http.go

```diff
@@ -143,30 +143,59 @@ func (df *dualFetcher) Do(req *http.Request) (*http.Response, error) {
 		}
 	}()
 
+	// Create channels for the final response or error. The response channel will be filled
+	// in the case of any successful response as well as a non-error response for the second
+	// response received. The error channel will only be filled if the first response non-successful
+	// and the second is an error.
+	finalResponseCh := make(chan *http.Response, 1)
+	finalErrorCh := make(chan error, 1)
+
+	go readResponses(finalResponseCh, responses, finalErrorCh, errs)
+
+	for {
+		select {
+		case resp := <-finalResponseCh:
+			return resp, nil
+		case err := <-finalErrorCh:
+			return nil, err
+		}
+	}
+}
+
+func readResponses(finalResponse chan *http.Response, responses chan *http.Response, finalErr chan error, errs chan error) {
 	for i := 0; i < 2; i++ {
 		select {
 		case resp := <-responses:
 			if i == 1 {
-				log.Debugf("Got second response -- sending")
-				return resp, nil
+				log.Debug("Got second response -- sending")
+				finalResponse <- resp
 			} else if success(resp) {
-				log.Debugf("Got good response")
-				// Returning preemptively here means the second response
-				// will not be closed properly. We need to ultimately
-				// handle that.
-				return resp, nil
+				log.Debug("Got good response")
+				finalResponse <- resp
```

Comment:
Inspired by a golang.org blog article, an alternative way would be to spawn a goroutine to receive response from channel and close the body, so we can return directly here instead of sending to channel.

Current method is good enough, don't bother change it.